standalone/prb/src/worker.rs

```diff
@@ -0,0 +1,1100 @@
+        let pr = crate::pruntime::create_client(w.endpoint.clone());
+        let pr = Arc::new(pr);
+        let (tx, rx) = mpsc::unbounded_channel::<WorkerLifecycleCommand>();
+
+        let mut ret = Self {
+            id: w.id.clone(),
+            self_ref: None,
+            sm_tx: None,
+            worker: w,
+            state: WorkerLifecycleState::Starting,
+            tx,
+            rx: Arc::new(TokioMutex::new(rx)),
+            ctx,
+            pr,
+            info: None,
+            last_message: String::new(),
+            session_info: None,
+        };
+        ret.set_last_message("Starting lifecycle...");
+        Ok(ret)
+    }
+
+    pub fn set_last_message(&mut self, m: &str) {
+        let time: DateTime<Local> = Local::now();
+        let worker = &self.worker;
+        self.last_message = format!("[{time}] {m}");
+        info!(
+            "Worker {}({}, {}): {}",
+            &worker.name, &worker.id, &worker.endpoint, &m
+        );
+    }
+
+    pub async fn start(c: WrappedWorkerContext) {
+        debug!("WorkerContext::start");
+        let cc = c.clone();
+        let cc = cc.read().await;
+        let ctx = cc.ctx.clone();
+        let lm = use_lm_with_ctx!(ctx);
+
+        let worker = cc.worker.clone();
+        drop(ctx);
+        drop(cc);
+        if let WorkerManagerMessage::ResponseErr(err_str) = lm
+            .clone()
+            .send_to_main_channel_and_wait_for_response(
+                WorkerManagerMessage::ShouldStartWorkerLifecycle(c.clone()),
+            )
+            .await
+            .unwrap_or_else(|e| panic!("Failed to start worker {}: {e}", worker.name))
+        {
+            let cc = c.clone();
+            let mut cc = cc.write().await;
+            cc.state = WorkerLifecycleState::HasError(err_str);
+            drop(cc);
+        }
+
+        let _ = join(
+            tokio::spawn(Self::do_start(c.clone())),
+            Self::message_loop(c.clone()),
+        )
+        .await;
+    }
+
+    async fn restart(_c: WrappedWorkerContext) {}
+
+    async fn do_start(c: WrappedWorkerContext) {
+        let (tx, rx) = mpsc::unbounded_channel::<WorkerLifecycleState>();
+        let mut cc = c.write().await;
+        cc.sm_tx = Some(tx.clone());
+        drop(cc);
+        let _ = join(
+            Self::lifecycle_loop(c.clone(), rx),
+            Self::update_info_loop(c.clone()),
+        )
+        .await;
+    }
+
+    async fn lifecycle_loop(c: WrappedWorkerContext, mut sm_rx: WorkerLifecycleStateRx) {
+        let (lm, worker, _pr, sm_tx) = extract_essential_values!(c);
+
+        let _ = sm_tx.clone().send(WorkerLifecycleState::Starting);
+
+        while let Some(s) = sm_rx.recv().await {
+            let mut cc = c.write().await;
+            cc.state = s.clone();
+            drop(cc);
+
+            match s {
+                WorkerLifecycleState::Starting => {
+                    lifecycle_loop_state_handle_error!(Self::handle_on_starting, c);
```

Comment:
For readability, we should:
 - Consider turning the macro into functions
 - Or if it is sometimes tedious to make them into functions. In this case, if the macro is only used in this scope, consider defining the macro as near its caller site as possible. The `lifecycle_loop_state_handle_error` was defined hundreds of lines away which would make it hard to read.